### PR TITLE
Allow expedited blocks only from expedited nodes

### PR DESCRIPTION
**Pull request by ptschip**

### CODE REVIEW

File: src/parallel.cpp [hunk 1]
```diff
@@ -553,13 +553,16 @@ void HandleBlockMessageThread(CNode *pfrom, const string &strCommand, const CBlo
         int nTotalThinBlocksInFlight = 0;
         {
             LOCK(cs_vNodes);
+            // Erase this thinblock from the tracking map now that we're done with it.
+            if (pfrom->mapThinBlocksInFlight.erase(inv.hash))
```
Comment (kyuupichan) [inline src/parallel.cpp]: Does this need the map lock?

File: src/expedited.cpp [hunk 2]
```diff
@@ -202,6 +202,14 @@ bool HandleExpeditedBlock(CDataStream &vRecv, CNode *pfrom)
         uint256 blkHash = thinBlock.header.GetHash();
         CInv inv(MSG_BLOCK, blkHash);
 
+        // Message consistency checking
+        if (!IsThinBlockValid(pfrom, thinBlock.vMissingTx, thinBlock.header))
+        {
+            LOCK(cs_main);
+            Misbehaving(pfrom->GetId(), 100);
+            return error("Invalid EXPEDITED_MSG_XTHIN received");
+        }
```
Comment (kyuupichan) [inline src/expedited.cpp:223]: I would feel much better if the validity checks for an xthin block and expedited block were the same code.  Expedited is missing much of it.  One is just an unrequested form of the other, so apart from minor details their handling should be identical.

File: src/main.cpp [hunk 3]
```diff
@@ -6062,23 +6068,41 @@ bool ProcessMessage(CNode* pfrom, std::string strCommand, CDataStream& vRecv, in
...
 
-    else if (strCommand == NetMsgType::XPEDITEDREQUEST)
+    else if (strCommand == NetMsgType::XPEDITEDREQUEST && IsThinBlocksEnabled())
     {
+        if (!pfrom->ThinBlockCapable())
+        {
+            LOCK(cs_main);
+            Misbehaving(pfrom->GetId(), 100);
+            return error("XPEDITEDREQUEST message received from a non thinblock node, peer=%d", pfrom->GetId());
+        }
```
Comment (kyuupichan) [inline src/main.cpp]: I continue to believe we're banning without good cause.  This is easily detected and not harmful, a low misbehaving seems more appropriate in this and several other cases.  Can we come up with a policy about what deserves a ban and apply it consistently?

File: src/test/exploit_tests.cpp [hunk 4]
```diff
@@ -84,191 +84,7 @@ uint256 nullhash;
...
-    BOOST_CHECK(dummyNode1.fBUVersionSent);
-    BOOST_CHECK(!CNode::IsBanned(addr1));
-
-    // Receive BUVERACK but no BUVERSION sent
-    dummyNode1.fSuccessfullyConnected = true;
-    dummyNode1.fBUVersionSent = false; // should cause ban
-    ProcessMessage(&dummyNode1, NetMsgType::BUVERACK, vRecv1, GetTime());
-    SendMessages(&dummyNode1);
-    BOOST_CHECK(!dummyNode1.fBUVersionSent);
-    BOOST_CHECK(CNode::IsBanned(addr1));
```
Comment (sickpig) [inline src/test/exploit_tests.cpp]: the removal of this test has already being merged in via  #490, right?

Comment (ptschip) [reply src/test/exploit_tests.cpp]: yes, i'll rebase a squash this out.

File: src/parallel.cpp (hunk 1 above)
Comment (ptschip) [reply src/parallel.cpp]: i'll update that...thanks.

File: src/main.cpp (hunk 3 above)
Comment (ftrader) [reply src/main.cpp]: I thought about the idea of such a policy, but didn't come up with any good solution except to move this policy into the hands of the node operator in the form of tweakable config:

https://bitco.in/forum/threads/buip053-tweakable-client-dos-responses.2074/

**Comment by ptschip** [comment]
@gandrewstone just fixed the merge conflict from the SENDHEADERS PR455

File: src/expedited.cpp [hunk 5]
```diff
@@ -286,3 +290,14 @@ void SendExpeditedBlock(const CBlock &block, const CNode *skip)
...
+bool IsExpeditedNode(const CNode *pfrom)
+{
+    // Check if this node is an expedited node
+    BOOST_FOREACH (const std::string &strAddrNode, mapMultiArgs["-expeditedblock"])
+    {
+        if (pfrom->addrName == strAddrNode)
+            return true;
+    }
+    return false;
+}
```
Comment (gandrewstone) [inline src/expedited.cpp]: expedited partners can be set via RPC call and we'll want an automatic option in the future.  So to discover whether a node is expedited, please use one of these global vars.
std::vector<CNode*> xpeditedBlk; // (256,(CNode*)NULL);    // Who requested expedited blocks from us
std::vector<CNode*> xpeditedBlkUp; //(256,(CNode*)NULL);  // Who we requested expedited blocks from
std::vector<CNode*> xpeditedTxn; // (256,(CNode*)NULL);

File: src/main.cpp (hunk 3 above)
Comment (gandrewstone) [reply src/main.cpp]: let's be more tolerant here.  How about Misbehaving 1 just to stop lots of repeat requests.  Also let's keep all the logic in HandleExpeditedRequest().  I think that this message handling function is too large and should be refactored to look simply like:
if (strCommand == NetMsgType::MSG1) HandleMSG1()
else if (strCommand == NetMsgType::MSG2) HandleMSG2()
...

and this HandleExpeditedRequest() is starting to go there.  Make it return a boolean if we need an early return from the message handling.

File: src/main.cpp [hunk 6]
```diff
@@ -5921,23 +5927,41 @@ bool ProcessMessage(CNode* pfrom, std::string strCommand, CDataStream& vRecv, in
...
+        if (!pfrom->ThinBlockCapable())
+        {
+            LOCK(cs_main);
+            Misbehaving(pfrom->GetId(), 100);
+            return error("XPEDITEDBLK message received from a non thinblock node, peer=%d", pfrom->GetId());
+        }
+
+        // ignore the expedited message unless we are near the chain tip...
         if (!fImporting && !fReindex && IsChainNearlySyncd())
         {
```
Comment (gandrewstone) [inline src/main.cpp]: similarly, let's put all this checking logic in HandleExpeditedBlock()

File: src/main.cpp [hunk 7]
```diff
@@ -5921,23 +5927,41 @@ bool ProcessMessage(CNode* pfrom, std::string strCommand, CDataStream& vRecv, in
     }
 
 
-    else if (strCommand == NetMsgType::XPEDITEDREQUEST)
+    else if (strCommand == NetMsgType::XPEDITEDREQUEST && IsThinBlocksEnabled())
```
Comment (gandrewstone) [inline src/main.cpp]: If conditions like (&& IsThinBlocksEnabled()) are added to these top level if statements, then the code will fall through to the final else that issues this log:
        // Ignore unknown commands for extensibility
        LogPrint("net", "Unknown command \"%s\" from peer=%d\n", SanitizeString(strCommand), pfrom->id);

I don't think we want that log printed if !IsThinBlocksEnabled().  Let's put the IsThinBlocksEnabled() in HandleExpeditedRequest()

File: src/main.cpp [hunk 8]
```diff
@@ -5921,23 +5927,41 @@ bool ProcessMessage(CNode* pfrom, std::string strCommand, CDataStream& vRecv, in
...
+            Misbehaving(pfrom->GetId(), 100);
+            return error("XPEDITEDREQUEST message received from a non thinblock node, peer=%d", pfrom->GetId());
+        }
+
         HandleExpeditedRequest(vRecv, pfrom);
     }
-    else if (strCommand == NetMsgType::XPEDITEDBLK)
+
+
+    else if (strCommand == NetMsgType::XPEDITEDBLK && IsThinBlocksEnabled() && IsExpeditedNode(pfrom))
```
Comment (gandrewstone) [inline src/main.cpp]: As noted earlier, if additional conditions like (&& IsThinBlocksEnabled()) are added to these top level if statements, then the code will fall through to the final else that issues an "Unknown command" log.

File: src/expedited.cpp (hunk 5 above)
Comment (ptschip) [reply src/expedited.cpp]: thanks i'll update...that's much better.

**Comment by ptschip** [comment]
@gandrewstone merging your pull but still sendheaders.py is broken...I'm going to have to take out the bans on unrequested blocks, even though they are slight ones.  There is no risk IMO, it's mainly a nuisance that some nodes send us unrequested blocks.  The problem here is that mininode has issues with not being able to send blocks unrequested and since it will likely take some time to figure out a good approach to testing I think the best thing is to remove these misbehaviors for the time being.

**Pull request merged**